\section{Material \& Methods}
Overview of method for analysis (this can roughly be paraphrased for the
purpose of manuscripts and grants):
\vspace{1.5cm}

\noindent
All sequence reads were processed by the NGS analysis pipeline of the SILVA
rRNA gene database project (SILVAngs 1.3) \citep{quast2013}.
Each read was aligned using the SILVA Incremental Aligner (SINA SINA v1.2.10 for ARB SVN (revision 21008))
\citep{pruesse2012} against the SILVA SSU rRNA SEED and quality
controlled \citep{quast2013}. 
Reads shorter than 200 aligned nucleotides and reads with more
than 2\% of ambiguities, or 2\% of homopolymers,
respectively, were excluded from further processing.
Putative contaminations and artefacts, reads with a low alignment quality
(50  alignment identity, 40 alignment score reported by
SINA), were idendified and excluded from downstream analysis.

After these inital steps of quality control, identical reads were
identified (dereplication), the unique reads were clustered (OTUs), 
on a per sample basis, and the reference read of each OTU was classified.
Dereplication and clustering was done using cd-hit-est (version
3.1.2; \url{http://www.bioinformatics.org/cd-hit}) \citep{li2006}
running in \emph{accurate mode}, ignoring overhangs, and applying identity
criteria of 1.00 and 0.98, respectively. 
The classification was performed by a local nucleotide BLAST search against
the non-redundant version of the SILVA SSU Ref dataset (release
128; \url{http://www.arb-silva.de}) using blastn (version
2.2.30+; \url{http://blast.ncbi.nlm.nih.gov/Blast.cgi}) with standard
settings \citep{camacho2009}.

The classification of each OTU reference read was mapped onto all reads
that were assigned to the respective OTU.
This yields quantative information (number of individual reads per taxonomic
path), within the limitations of PCR and sequencing technique biases, as 
well as, multiple rRNA operons.
Reads without any BLAST hits or reads with weak BLAST hits, where the
function ``(\% sequence identity + \% alignment coverage)/2'' did not exceed
the value of 93, remain unclassified.
These reads were assigned to the meta group ``No Relative'' in the
SILVAngs fingerprint and Krona charts \citep{ondov2011}.

This method was first used in the publications of \cite{klindworth2013} and
\cite{ionescu2012} 
